onnxruntime/core/framework/kernel_registry_manager.h

```diff
@@ -17,46 +18,56 @@ class KernelRegistry;
 class OpKernel;
 class SessionState;
 
-enum class KernelRegistryPriority {
-  HighPriority,
-  LowPriority
-};
-
 // Kernel registries' manager.
 // There're 2 kinds of kernel registries with priority from high to low as below,
 // 1. Custom execution provider type specific kernel registries.
 // 2. common execution provider type specific kernel registries.
 // The 1st and 2nd ones are shared across sessions.
+
+//This class is thread safe
 class KernelRegistryManager {
  public:
   KernelRegistryManager() = default;
 
-  void RegisterKernels(const ExecutionProviders& execution_providers,
-                       KernelRegistryPriority priority = KernelRegistryPriority::LowPriority);
+  //Register kernels from providers
+  void RegisterKernels(const ExecutionProviders& execution_providers);
 
-  void RegisterKernelRegistry(std::shared_ptr<KernelRegistry> kernel_registry, KernelRegistryPriority priority);
+  // The registry passed in this function has highest priority than anything already in this KernelRegistryManager,
+  // and anything registered from RegisterKernels
+  // For example, if you do:
+  // RegisterKernels(providers)
+  // RegisterKernelRegistry(A);
+  // RegisterKernelRegistry(B);
+  // Then B > A > providers
+  void RegisterKernelRegistry(std::shared_ptr<KernelRegistry> kernel_registry);
 
   Status CreateKernel(const onnxruntime::Node& node,
                       const IExecutionProvider& execution_provider,
                       const SessionState& session_state,
                       /*out*/ std::unique_ptr<OpKernel>& op_kernel) const;
 
+  //This function assumes the node is already assigned to an execution provider
+  //Don't call this function before graph partition is done
   Status SearchKernelRegistry(const onnxruntime::Node& node,
                               /*out*/ const KernelCreateInfo** kernel_create_info) const;
 
+  bool HasImplementationOf(const Node& node, const std::string& provider_type) const;
+
   // Get all kernel registries. There are no nullptr entries.
-  std::vector<const KernelRegistry*> GetAllKernelRegistries() const {
+  std::vector<const KernelRegistry*> GetKernelRegistriesByProviderType(const std::string& type) const {
     std::vector<const KernelRegistry*> result;
     for (auto& registry : kernel_registries_) {
       result.push_back(registry.get());
     }
+    auto iter = provider_type_to_registry_.find(type);
+    if (iter != provider_type_to_registry_.end())
+      result.push_back(iter->second.get());
     return result;
   }
-
- private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(KernelRegistryManager);
 
```

Comment:
nit: prefer whitespace separating unrelated code. Also does ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE need to move to the 'public' section?

Reply:
"Deleted member function should be public"
See: 

1. https://abseil.io/tips/143
2. https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types
"deleting the appropriate operations in the **public** section of the declaration"

Reply:
Cool. Thanks for the pointers - learnt something. I was just following the previous usages of doing that in the 'private' section. Good to know why doing it in 'public' is better.